docs/blog/posts/write-code-to-refactor.md

```diff
@@ -0,0 +1,402 @@
+
+```python
+def format_data(data):
+  headers = extract_headers(data)
+	data = format_dataframe(data,headers)
+	upload_to_database(data)
+  return data
+
+def process():
+	data = get_data()
+	data = format_data(data)
+```
+
+There are two main problems here
+
+1. We're introducing a strange side effect that `format_data` doesn't reflect
+
+2. We're mutating state, which makes it difficult to debug issues when our logic gets more complicated
+
+A better way to write this logic would include splitting up the saving of data and the formatting of data into two separate functions. That way our functions closely reflect the actual actions that are being performed. We'd also try to minimize the mutation of state so that it's easier to trace errors when they happen
+
+```python
+def format_data(data):
+  headers = extract_headers(data)
+	return format_dataframe(data,headers)
+
+def save_data(data):
+  upload_to_database(data)
+
+def process():
+	data = get_data()
+	formatted_data = format_data(data)
+  save_to_database(data)
+```
+
+#### Be consistent with abstraction
+
+It's much nicer to keep code at the same level of abstraction. In the example above, we can see two kinds of abstraction levels
+
+- `Implementation Detail` : An example of something like this might be a function with a complex if-else logic 
+- `Action Detail` : These tend to be functions with verbs as names - eg. `save_data`, `send_request`
+
+```python
+HUGGING_FACE=[
+  "BAAI/bge-small-en-v1.5",
+]
+
+OPENAI = [
+  "text-embedding-3-small"
+]
+
+def inference(model_name,data):
+  if model_name in HUGGING_FACE:
+    model = EmbeddingModel.from_hf(model_name)
+  elif model_name in OPENAI:
+    model = EmbeddingModel.from_openai(model_name)
+  else:
+    raise ValueError
+  
+	embeddings = model.infer(data)
+```
+
+We're trying to handle the complex initialisation of the model along with the abstract `.infer` method that the `EmbeddingModel` class provides. As a result, the code looks clunky and messy. A better way to write this would be to have a separate function to handle the initialisation of the model itself
+
+```python
+HUGGING_FACE=[
+  "BAAI/bge-small-en-v1.5",
+]
+
+OPENAI = [
+  "text-embedding-3-small"
+]
+
+def initialise_model(model_name:str):
+	if model_name in HUGGING_FACE:
+    model = EmbeddingModel.from_hf(model_name)
+  elif model_name in OPENAI:
+    model = EmbeddingModel.from_openai(model_name)
+  else:
+    raise ValueError(f"Invalid model name of {model_name} was provided")
+
+def inference(model_name:str,data:dict):
+  model = initialise_model(model_name)
+	embeddings = model.infer(data)
+```
+
+Alternatively, we could take advantage of the `@classmethod` decorator again and define a method on the class which takes in a model_name and sets the relevant parameters.
+
+## Training Models
```

Comment:
This should almost be a second post. 

Basically saying that we should just get things wired up first and then increase data size.